### PR TITLE
feat(events,applicationautoscaling): schedule can be a token

### PR DESCRIPTION
**Pull request by sneharathod**
Fix for Issue: https://github.com/aws/aws-cdk/issues/9413

Testing:
Unit tests added for Schedule.rate used in ApplicationAutoScaling and Events.

* [X] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [X] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [X] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parents, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts [hunk 1]
```diff
@@ -26,4 +26,22 @@ export = {
...
     test.done();
   },
+
+  'rate can be token'(test: Test) {
+    const stack = new Stack();
+    const invokeScheduleInMinutes = new CfnParameter(stack, 'invokeScheduleInMinutes', {
+      type: 'Number',
+      description: 'The minutes between scheduled invocations.',
+      default: 5,
+    });
```
Comment (nija-at) [inline packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts:36]: If you switch to using the `Lazy` class (which also uses the same token mechanism), it would be easier to test. Something like this -

```ts
const lazyDuration = Duration.minutes(Lazy.number({ produce: () => 5 }));
const rate = appscaling.Schedule.rate(lazyDuration)
test.equal('rate(<token> minutes)', rate);
test.equal('rate(5 minutes)', stack.resolve(rate));

File: packages/@aws-cdk/aws-events/lib/schedule.ts [hunk 2]
```diff
@@ -16,7 +16,11 @@ export abstract class Schedule {
   /**
    * Construct a schedule from an interval and a time unit
    */
-  public static rate(duration: Duration): Schedule {
+  public static rate(duration: Duration): Schedule {    
+    if(duration.isUnresolved()){
+      return new LiteralSchedule(`rate(${duration.toHumanString()})`);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-events/lib/schedule.ts:21]: I believe this [only supports minutes, hours and days](https://docs.aws.amazon.com/eventbridge/latest/userguide/scheduled-events.html#rate-expressions) but `Duration` also supports seconds and milliseconds. It could support more in the future.

Also, I believe the number must not be a fraction

Could we validations for these, and throw an appropriate error.

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts [hunk 3]
```diff
@@ -26,4 +26,22 @@ export = {
...
+    test.equal('rate(<token> minutes)', appscaling.Schedule.rate(Duration.minutes(invokeScheduleInMinutes.valueAsNumber))
+      .expressionString);
+    test.done();
+  },
+
+  'rate can be in minutes'(test: Test) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts:42]: curious: why do we need this extra test case?

File: packages/@aws-cdk/aws-events/test/test.input.ts [hunk 4]
```diff
@@ -102,6 +102,49 @@ export = {
...
+        schedule: Schedule.rate(cdk.Duration.minutes(invokeScheduleInMinutes.valueAsNumber)),
+      });
+      const user = new User(stack, 'User');
+
+      // WHEN
+      rule.addTarget(new SomeTarget(RuleTargetInput.fromObject({ userArn: user.userArn })));
+
+      // THEN
+      expect(stack).to(haveResourceLike('AWS::Events::Rule', {
+        Targets: [
```
Comment (nija-at) [inline packages/@aws-cdk/aws-events/test/test.input.ts:126]: We should be asserting the `ScheduleExpression` property instead of the `Targets` property, no?

File: packages/@aws-cdk/aws-events/test/test.input.ts [hunk 5]
```diff
@@ -102,6 +102,49 @@ export = {
       }));
       test.done();
     },
+
+    'can use token from cfnParameter'(test: Test) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-events/test/test.input.ts:106]: This feels like a duplicate of the test in `test.rule.ts`. Does this check anything additional that the other test case doesn't?

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.scalable-target.ts [hunk 6]
```diff
@@ -85,6 +85,40 @@ export = {
     test.done();
   },
 
+  'add scheduled scaling with cfnParameter'(test: Test) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-applicationautoscaling/test/test.scalable-target.ts:88]: same as my other comment. Feels like the test in test.cron.ts should provide sufficient cover. Do we need an extra test case?

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for submitting this PR 😊 My comments are below.

Looks like the build and other PR checks are failing. Can you look into them?

As for the failing CodeBuild job, the build logs are available as a comment above in this PR.

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts (hunk 3 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts:42]: I wanted to make sure we had something around regular minutes.

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts (hunk 1 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts:36]: Sure, will give it a try.

File: packages/@aws-cdk/aws-events/lib/schedule.ts (hunk 2 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-events/lib/schedule.ts:21]: Sure. I see that Seconds should work with Schedule rate since duration.toSeconds() has convert() which would match the unit and not convert if units match. But, milliseconds is left. I can add a check for it being unsupported.Does the Error "Unsupported duration format - millisecond" sound ok? 

Regarding fraction, I saw convert() has a check
```
 if (!Number.isInteger(value) && integral) {
    throw new Error(`'${amount} ${fromUnit}' cannot be converted into a whole number of ${toUnit}.`);
  } 
``` 
Does it help?

File: packages/@aws-cdk/aws-events/test/test.input.ts (hunk 4 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-events/test/test.input.ts:126]: Right, my bad.

File: packages/@aws-cdk/aws-events/test/test.input.ts (hunk 5 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-events/test/test.input.ts:106]: Let me check and remove if it is duplicate.

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.scalable-target.ts (hunk 6 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-applicationautoscaling/test/test.scalable-target.ts:88]: Sure, let me check and remove if not required.

File: packages/@aws-cdk/aws-events/lib/schedule.ts (hunk 2 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-events/lib/schedule.ts:21]: Regarding seconds, I see in the doc it only mentions minute | minutes | hour | hours | day | days. Let me add check for seconds too.

File: packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts (hunk 1 above)
Comment (sneharathod) [reply packages/@aws-cdk/aws-applicationautoscaling/test/test.cron.ts:36]: @nija-at Are you suggesting me to modify in all the testcases where I have used cfnParameters or few?

**Comment by nija-at** [comment]
I missed this in the first round but `toHumanString()` is not going to work here. It's because it places '<token>' when it finds an unresolved token.

What should happen for this to work is that the actual token encoded number - https://docs.aws.amazon.com/cdk/latest/guide/tokens.html#tokens_number - must be produced here.

So you will likely have to introduce a new API, say `format()`, that would do something like this -

```ts
const duration = Duration.seconds(5);
const output = duration.format('duration is ${amount} ${unit}');

output <-- "duration is 5 seconds"
```

With tokens, it would be  -

```ts
const duration = Duration.seconds(Lazy.number({ produce: () => 5 }));
const output = duration.format('duration is ${amount} ${unit}');

output <-- "duration is -1.8881545897087626e+289 seconds" // this is a number encoded token
stack.resolve(output) <-- "duration is 5 seconds"
```

**Comment by sneharathod** [comment]
I had tried sending token value as it is like -1.8881545897087626e+289. I
had modified toString() in duration.ts and used it instead of humanstring.

But it had not worked too with stack.resolve. Will check again.

**Comment by sneharathod** [comment]
First time user issues! I am seeing some 80+ merge commits since I pulled from master aws-cdk into my fork repo. Let me know if we could discard this one and create new PR from another fork repository.

**Comment by nija-at** [comment]
@sneharathod - this is quite normal in the CDK repo as we have quite a few commits going in each day. 

You can choose one of two things - (a) do nothing; when the PR is approved, our bot will automatically update your branch to track the latest and then merge, or (b) click on the 'Update branch' button at the bottom of the PR that will update your branch to track the latest commit on `master`.

It is necessary to perform (b) explicitly only if there are merge conflicts with master that need to be resolved manually. You will notified of this in the area that currently says "This branch is out-of-date with the base branch".
Your PR currently has no merge conflicts.

File: packages/@aws-cdk/aws-applicationautoscaling/lib/schedule.ts [hunk 7]
```diff
@@ -17,6 +17,13 @@ export abstract class Schedule {
    * Construct a schedule from an interval and a time unit
    */
   public static rate(duration: Duration): Schedule {
+    const validDurationUnit = ['minute', 'minutes', 'hour', 'hours', 'day', 'days'];
+    if (validDurationUnit.indexOf(duration.unitLabel()) === -1) {
+      throw new Error('Allowed unit for scheduling is: \'minute\',\'minutes\', \'hour\', \'hours\', \'day\' or \'days\'');
```
Comment (nija-at) [inline packages/@aws-cdk/aws-applicationautoscaling/lib/schedule.ts:22]: Some code readability.

```suggestion
    if (!validDurationUnit.includes(duration.unitLabel())) {
      throw new Error("Allowed unit for scheduling is: 'minute', 'minutes', 'hour', 'hours', 'day' or 'days'");
```

File: packages/@aws-cdk/aws-events/README.md [hunk 8]
```diff
@@ -83,6 +83,7 @@ onCommitRule.addTarget(new targets.SnsTopic(topic, {
 ## Scheduling
 
 You can configure a Rule to run on a schedule (cron or rate).
+Rate can be specified in Minutes/Hours/Days.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-events/README.md:86]:
```suggestion
Rate must be specified in minutes, hours or days.
```

File: packages/@aws-cdk/cfnspec/spec-source/000_CloudFormationResourceSpecification.json [hunk 9]
```diff
@@ -66658,6 +66658,12 @@
           "Required": false,
           "UpdateType": "Immutable"
         },
+        "EnableExecuteCommand": {
```
Comment (nija-at) [inline packages/@aws-cdk/cfnspec/spec-source/000_CloudFormationResourceSpecification.json:66661]: can you take a look at why this is coming in? Seems unrelated to your change.

File: yarn.lock [hunk 10]
```diff
@@ -3687,16 +3682,6 @@ dot-prop@^5.1.0:
   dependencies:
     is-obj "^2.0.0"
 
-dotenv-json@^1.0.0:
```
Comment (nija-at) [inline yarn.lock:3690]: You changes should not impact this file either.

File: packages/@aws-cdk/aws-events/test/test.rule.ts [hunk 11]
```diff
@@ -49,6 +49,47 @@ export = {
...
+        'MyScheduledRuleD15B5BA3': {
+          'Type': 'AWS::Events::Rule',
+          'Properties': {
+            'Name': 'rateInMinutes',
+            'ScheduleExpression': 'rate(5 minutes)',
+            'State': 'ENABLED',
+          },
+        },
+      },
+    });
```
Comment (nija-at) [inline packages/@aws-cdk/aws-events/test/test.rule.ts:74]: When asserting properties of single resources, we use -

```suggestion
    expect(stack).toHaveResourceLike('AWS::Events::Rule', {
      'Name': 'rateInMinutes',
      'ScheduleExpression': 'rate(5 minutes)',
    });
```

File: packages/@aws-cdk/core/lib/duration.ts [hunk 12]
```diff
@@ -252,6 +252,28 @@ export class Duration {
     }
     return ret;
   }
+
+  /**
```
Comment (nija-at) [inline packages/@aws-cdk/core/lib/duration.ts:256]: Can you add unit tests for the new methods in this file?

**Review by nija-at — DISMISSED** [review]
Much better and I like the approach. Thanks for the feedback. 

Some small suggestions below.

**Comment by sneharathod** [comment]
Sure. I missed them.

**Review by nija-at — CHANGES_REQUESTED** [review]
Setting this back to 'request changes' to address remaining comments.

**Review by nija-at — APPROVED** [review]
